File: agent/bindings/kotlin/lib/src/main/kotlin/com/sourcegraph/cody/protocol_generated/WorkspaceEditOperation.kt
@file:Suppress("FunctionName", "ClassName", "unused", "EnumEntryName", "UnusedImport")
package com.sourcegraph.cody.protocol_generated

import com.google.gson.annotations.SerializedName
import com.google.gson.Gson
import com.google.gson.JsonDeserializationContext
import com.google.gson.JsonDeserializer
import com.google.gson.JsonElement
import java.lang.reflect.Type

sealed class WorkspaceEditOperation {
  companion object {
    val deserializer: JsonDeserializer<WorkspaceEditOperation> =
      JsonDeserializer { element: JsonElement, _: Type, context: JsonDeserializationContext ->
        when (element.asJsonObject.get("type").asString) {
          "create-file" -> context.deserialize<CreateFileOperation>(element, CreateFileOperation::class.java)
          "rename-file" -> context.deserialize<RenameFileOperation>(element, RenameFileOperation::class.java)
          "delete-file" -> context.deserialize<DeleteFileOperation>(element, DeleteFileOperation::class.java)
          "edit-file" -> context.deserialize<EditFileOperation>(element, EditFileOperation::class.java)
          else -> throw Exception("Unknown discriminator ${element}")
        }
      }
  }
}

data class CreateFileOperation(
  val type: TypeEnum? = null, // Oneof: create-file
  val uri: String? = null,
  val options: WriteFileOptions? = null,
  val textContents: String? = null,
  val metadata: WorkspaceEditEntryMetadata? = null,
) : WorkspaceEditOperation() {

  enum class TypeEnum {
    @SerializedName("create-file") `Create-file`,
  }
}

data class RenameFileOperation(
  val type: TypeEnum? = null, // Oneof: rename-file
  val oldUri: String? = null,
  val newUri: String? = null,
  val options: WriteFileOptions? = null,
  val metadata: WorkspaceEditEntryMetadata? = null,
) : WorkspaceEditOperation() {

  enum class TypeEnum {
    @SerializedName("rename-file") `Rename-file`,
  }
}

data class DeleteFileOperation(
  val type: TypeEnum? = null, // Oneof: delete-file
  val uri: String? = null,
  val deleteOptions: DeleteOptionsParams? = null,
  val metadata: WorkspaceEditEntryMetadata? = null,
) : WorkspaceEditOperation() {

  enum class TypeEnum {
    @SerializedName("delete-file") `Delete-file`,
  }
}

data class EditFileOperation(
  val type: TypeEnum? = null, // Oneof: edit-file
  val uri: String? = null,
  val edits: List<TextEdit>? = null,
) : WorkspaceEditOperation() {

  enum class TypeEnum {
    @SerializedName("edit-file") `Edit-file`,
  }
}

File: agent/bindings/csharp/WorkspaceEditOperation.cs
using System.Text.Json.Serialization;

namespace Cody.Core.Agent.Protocol
{

  [JsonConverter(typeof(WorkspaceEditOperationConverter))]
  public abstract class WorkspaceEditOperation
  {
      private class WorkspaceEditOperationConverter : JsonConverter<WorkspaceEditOperation>
      {
        public override WorkspaceEditOperation Read(ref Utf8JsonReader reader, Type typeToConvert, JsonSerializerOptions options)
        {
          var jsonDoc = JsonDocument.ParseValue(ref reader);
          var discriminator = jsonDoc.RootElement.GetProperty("type").GetString();
          switch (discriminator)
          {
            case "create-file":
              return JsonSerializer.Deserialize<CreateFileOperation>(jsonDoc.RootElement.GetRawText(), options);
            case "rename-file":
              return JsonSerializer.Deserialize<RenameFileOperation>(jsonDoc.RootElement.GetRawText(), options);
            case "delete-file":
              return JsonSerializer.Deserialize<DeleteFileOperation>(jsonDoc.RootElement.GetRawText(), options);
            case "edit-file":
              return JsonSerializer.Deserialize<EditFileOperation>(jsonDoc.RootElement.GetRawText(), options);
            default:
              throw new JsonException($"Unknown discriminator {discriminator}");
            }
        }
        public override void Write(Utf8JsonWriter writer, ${name} value, JsonSerializerOptions options)
        {
          JsonSerializer.Serialize(writer, value, value.GetType(), options);
        }
  }

  public class CreateFileOperation : WorkspaceEditOperation
  {
    [JsonProperty(PropertyName = "type")]
    public string Type { get; set; } // Oneof: create-file
    [JsonProperty(PropertyName = "uri")]
    public string Uri { get; set; }
    [JsonProperty(PropertyName = "options")]
    public WriteFileOptions Options { get; set; }
    [JsonProperty(PropertyName = "textContents")]
    public string TextContents { get; set; }
    [JsonProperty(PropertyName = "metadata")]
    public WorkspaceEditEntryMetadata Metadata { get; set; }
  }

  public class RenameFileOperation : WorkspaceEditOperation
  {
    [JsonProperty(PropertyName = "type")]
    public string Type { get; set; } // Oneof: rename-file
    [JsonProperty(PropertyName = "oldUri")]
    public string OldUri { get; set; }
    [JsonProperty(PropertyName = "newUri")]
    public string NewUri { get; set; }
    [JsonProperty(PropertyName = "options")]
    public WriteFileOptions Options { get; set; }
    [JsonProperty(PropertyName = "metadata")]
    public WorkspaceEditEntryMetadata Metadata { get; set; }
  }

  public class DeleteFileOperation : WorkspaceEditOperation
  {
    [JsonProperty(PropertyName = "type")]
    public string Type { get; set; } // Oneof: delete-file
    [JsonProperty(PropertyName = "uri")]
    public string Uri { get; set; }
    [JsonProperty(PropertyName = "deleteOptions")]
    public DeleteOptionsParams DeleteOptions { get; set; }
    [JsonProperty(PropertyName = "metadata")]
    public WorkspaceEditEntryMetadata Metadata { get; set; }
  }

  public class EditFileOperation : WorkspaceEditOperation
  {
    [JsonProperty(PropertyName = "type")]
    public string Type { get; set; } // Oneof: edit-file
    [JsonProperty(PropertyName = "uri")]
    public string Uri { get; set; }
    [JsonProperty(PropertyName = "edits")]
    public TextEdit[] Edits { get; set; }
  }
  }
}
